core/src/main/scala/org/broadinstitute/dsde/rawls/fastpass/FastPassService.scala

```diff
@@ -218,115 +218,143 @@ class FastPassService(protected val ctx: RawlsRequestContext,
-                                            childWorkspace: Workspace
-  ): ReadWriteAction[Unit] = {
-    if (!config.enabled) {
-      logger.debug(s"FastPass is disabled. Will not grant FastPass access to ${parentWorkspace.toWorkspaceName}")
-      return DBIO.successful()
-    }
-
-    DBIO.from(quotaAvailableForClonedWorkspaceFastPassGrants(parentWorkspace, childWorkspace)).flatMap {
-      quotaAvailable =>
+    try
+      DBIO.from(quotaAvailableForNewWorkspaceFastPassGrants(workspace)).flatMap { quotaAvailable =>
         if (quotaAvailable) {
-          logger.info(
-            s"Adding FastPass access for ${ctx.userInfo.userEmail} in workspace being cloned ${parentWorkspace.toWorkspaceName}"
-          )
+          logger
+            .info(
+              s"Adding FastPass access for ${ctx.userInfo.userEmail.value} in workspace ${workspace.toWorkspaceName}"
+            )
           val expirationDate = DateTime.now(DateTimeZone.UTC).plus(config.grantPeriod.toMillis)
           for {
             maybeUserStatus <- DBIO.from(samDAO.getUserStatus(ctx))
             if maybeUserStatus.isDefined
             samUserInfo = maybeUserStatus.map(SamUserInfo.fromSamUserStatus).orNull
 
-            petEmail <- DBIO.from(samDAO.getUserPetServiceAccount(ctx, childWorkspace.googleProjectId))
-
+            roles <- DBIO
+              .from(samDAO.listUserRolesForResource(SamResourceTypeNames.workspace, workspace.workspaceId, ctx))
+            petEmail <- DBIO.from(samDAO.getUserPetServiceAccount(ctx, workspace.googleProjectId))
             userType = getUserType(samUserInfo.userEmail)
             userAndPet = UserAndPetEmails(samUserInfo.userEmail, userType, petEmail)
-            _ <- setupBucketRoles(parentWorkspace,
-                                  Set(SamWorkspaceRoles.reader),
-                                  userAndPet,
-                                  samUserInfo,
-                                  expirationDate
-            )
+            _ <- setupProjectRoles(workspace, roles, userAndPet, samUserInfo, expirationDate)
+            _ <- setupBucketRoles(workspace, roles, userAndPet, samUserInfo, expirationDate)
             _ <- DBIO
               .from(openTelemetry.incrementCounter("fastpass-granted-user", tags = openTelemetryTags).unsafeToFuture())
           } yield ()
         } else {
           logger.info(
-            s"Not enough IAM Policy Role Binding quota available to add FastPass access for ${ctx.userInfo.userEmail.value} in parent workspace ${parentWorkspace.toWorkspaceName}"
+            s"Not enough IAM Policy Role Binding quota available to add FastPass access for ${ctx.userInfo.userEmail.value} in workspace ${workspace.toWorkspaceName}"
           )
           DBIO.successful()
         }
+      }
+    catch {
+      case e: Exception =>
+        logger.error(s"Failed to add FastPasses for new user in ${workspace.toWorkspaceName}", e)
```

Comment:
Might be good to emit a counter so we can better telemetry on how often this is happening.